doc/modules/mod_event_pusher_rabbit.md

```diff
@@ -0,0 +1,193 @@
+  a RabbitMQ server and receives a confirmation
+  * `message_payload_size` - size of a messages (in bytes) that was published to
+  a RabbitMQ server
+
+> All the above metrics have a prefix which looks as follows:  
+> `mongooseim.<xmpp_host>.backends.mod_event_pusher_rabbit.<metric_name>`.
+> For example a proper metric name would look like:
+> `mongooseim.localhost.backends.mod_event_pusher_rabbit.connections_active`
+
+### Current status
+
+This module is still in an experimental phase.
+
+### Guarantees
+
+There are no guarantees. The current implementation uses "best effort" approach
+which means that we don't care if a message is delivered to a RabbitMQ server.
+If a message couldn't be delivered to the server for any reason the module
+just updates appropriate metrics and print some log messages.
+
+### Type of exchanges
+
+By default all the exchanges used are of type `topic`. Using topic exchanges
+gives a lot of flexibility when binding queues to such an exchange by using
+`#` and `*` in binding keys. But flexibility comes at the cost of performance -
+imagine a scenario where there are thousands of users and AMQP consumers use
+bindig keys for particular users which looks like `user_N@host.#`. In such
+case RabbitMQ has to go through all the users in order to find out where
+a message should be sent to. This operations is proved to be costly. In a load
+test with 100k users a delay caused by this operation was substantial (about an
+order of magnitue higher than compared to a load test with 60k users).
+
+If perfromance is a top priority go for `direct` exchanges. Using this type of
+exchanges is proved to work efficiently with 100k users. Keep in mind it gives
+up flexibility over perfromance.
+
+### Publisher confirms
+
+By default publisher confirmations are disabled. However, one-to-one
+confirmations can be enabled (see
+[*RabbitMQ connection setup*](../advanced-configuration/outgoing-connections.md#rabbitmq-connection-setup)
+section). When a worker sends a message to a RabbitMQ server it waits for a
+confirmation from the server before it starts to process next message. This
+approach allows to introduce backpressure on a RabbitMQ server connection cause
+the server can reject messages when it's overloaded. On the other hand it can
+cause performance degratation.
+
+### Worker selection strategy
+
+The module uses `mongoose_wpool` for managing worker processes  and `avaiable_worker`
+strategy is in use. Different strategies imply different behaviours of the system.
+
+#### Event messages queueing
+
+When `available_worker` strategy is set all the event messages are queued in
+single worker pool manager process state. When diffrenet strategy is set e.g
+`next_worker` those messages are placed in worker processes inboxes. There is no
+possiblity to change the worker strategy for now.
+
+#### Event messages ordering
+
+`available_worker` strategy does not ensure that user events will be delivered to
```

Comment:
Because we use `cast` in `mod_event_pusher_rabbit` so I can imagine that the worker which received the first message is processing it longer that the one which received the second message. The reason for that might be the first message is simply larger that the second one. Does it make sense?

Reply:
Could be if there are multiple workers used - there is a race between workers on delivering messages.
Strict ordering requires serialization of message delivery by employing one worker only.
So actually mode of queuing doesn't matter (if workers are not restarted, that is if they do not loose mbox), it is only a performance optimization, what matters is the number of configured workers.